esr_key_value_database_intrinsics.md

```diff
@@ -0,0 +1,264 @@
+|`iterator_oob`    | -2    | Iterator is out-of-bounds |
+
+### kv_it_create
+
+```c++
+uint32_t kv_it_create(uint64_t db, uint64_t contract, const char* prefix, uint32_t size);
+```
+
+`kv_it_create` creates an iterator. The returned handle:
+
+* Starts at 1.
+* Counts up by 1 at each call, assuming no destroyed iterators are available.
+* If destroyed iterators are available, the most-recently-destroyed one is reinitialized and returned.
+
+The prefix limits the range of keys that the iterator covers. If the prefix is empty, the iterator
+covers the entire range of keys belonging to a contract within the database ID.
+
+The newly-created iterator has `iterator_oob` status.
+
+A consensus parameter limits the number of available iterators. `kv_it_create` aborts the
+transaction if the contract exceeds this.
+
+### kv_it_destroy
+
+```c++
+void     kv_it_destroy(uint32_t itr);
+```
+
+This destroys an iterator. It aborts the transaction if `itr` wasn't returned by a call to
+`kv_it_create`, or if `itr` has already been destroyed.
+
+### kv_it_status
+
+```c++
+it_stat  kv_it_status(uint32_t itr);
+```
+
+This returns the status of an iterator. It aborts the transaction if `itr` wasn't returned
+by a call to `kv_it_create`, or if `itr` has been destroyed.
+
+### kv_it_compare
+
+```c++
+int      kv_it_compare(uint32_t itr_a, uint32_t itr_b);
+```
+
+Compare the position of two iterators. It aborts the transaction if either iterator wasn't
+returned by `kv_it_create` or was destroyed. It also aborts if the iterators are from different
+databases or from different contracts.
+
+| Return Value | Description |
+|--------------|-------------|
+| -1           | `itr_a`'s key is less than `itr_b`'s key |
+| 0            | `itr_a`'s key is the same as `itr_b`'s key |
+| 1            | `itr_a`'s key is greater than `itr_b`'s key |
+
+If an iterator has status `iterator_erased`, then `kv_it_compare` uses the value of the erased key
+during comparison.
+
+If an iterator has status `iterator_oob`, then it compares greater than iterators with status
+`iterator_ok` or `iterator_erased`. Two iterators with status `iterator_oob` compare equal
+(0 return value).
+
+### kv_it_key_compare
+
+```c++
+int      kv_it_key_compare(uint32_t itr, const char* key, uint32_t size);
+```
+
+Compare the position of an iterator to a provided key. It aborts the transaction if `itr` wasn't
+returned by `kv_it_create` or was destroyed.
+
+| Return Value | Description |
+|--------------|-------------|
+| -1           | `itr`'s key is less than `key` |
+| 0            | `itr`'s key is the same as `key` |
+| 1            | `itr`'s key is greater than `key` |
+
+If `itr` has status `iterator_erased`, then `kv_it_key_compare` uses the value of the erased key
+during comparison. If `itr` has status `iterator_oob`, then it compares greater than 
+`key`, no matter what value `key` has.
+
+### kv_it_move_to_oob
+
+```c++
+it_stat  kv_it_move_to_oob(uint32_t itr);
```

Comment:
What is the point of the return value if it will always return `iterator_oob`?

Reply:
Consistency with the other functions

Reply:
Okay, that's fine.

Reply:
Okay, that's fine.

Reply:
Okay, that's fine.